node/src/types/chainspec/protocol_config.rs

```diff
@@ -0,0 +1,95 @@
+use datasize::DataSize;
+#[cfg(test)]
+use rand::Rng;
+use semver::Version;
+use serde::{Deserialize, Serialize};
+
+use casper_types::bytesrepr::{self, FromBytes, ToBytes};
+
+use crate::components::consensus::EraId;
+#[cfg(test)]
+use crate::testing::TestRng;
+
+/// The era whose end will trigger the upgrade process.
+#[derive(Copy, Clone, DataSize, PartialEq, Eq, Serialize, Deserialize, Debug)]
+pub(crate) struct ActivationPoint {
+    pub(crate) era_id: EraId,
+}
+
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize, DataSize, Debug)]
+pub struct ProtocolConfig {
+    #[data_size(skip)]
+    pub(crate) version: Version,
+    /// This protocol config applies to the era begun immediately after the activation point.
+    pub(crate) activation_point: ActivationPoint,
+}
+
+#[cfg(test)]
+impl ProtocolConfig {
+    /// Generates a random instance using a `TestRng`.
+    pub fn random(rng: &mut TestRng) -> Self {
+        let protocol_version = Version::new(
+            rng.gen_range(0, 10),
+            rng.gen::<u8>() as u64,
+            rng.gen::<u8>() as u64,
+        );
+        let activation_point = ActivationPoint {
+            era_id: EraId(rng.gen::<u8>() as u64),
+        };
+
+        ProtocolConfig {
+            version: protocol_version,
+            activation_point,
+        }
+    }
+}
+
+impl ToBytes for ProtocolConfig {
```

Comment:
Why do we need both `ToBytes/FromBytes` and `Serialized/Deserialize` impls?

Reply:
We need the `ToBytes` in order to use it when serializing the chainspec in order to generate its hash (we were using `bincode` before, but we should prefer our own format for this).  The `FromBytes` isn't strictly necessary I think, but I'd certainly want it at least for `#[cfg(test)]` so we can do a round trip test.

The serde traits allow us to easily encode/decode to/from TOML format.

Reply:
I realized that only later, after going through the first commit completely. Thanks :)